keps/sig-scheduling/3633-matchlabelselectors-to-podaffinity/README.md

```diff
@@ -263,10 +264,51 @@ metadata:
             values:
             - database
         topologyKey: topology.kubernetes.io/zone
-        matchlabelKeys: # ADDED
-        - pod-template-hash
+        matchLabelSelectors: # ADDED
+        - matchLabelKeys: pod-template-hash
+          operator: In
 ```
 
+#### Story 2
+
+When users want to achieve exclusive 1:1 job to domain placement in [kubernetes-sigs/jobset](https://github.com/kubernetes-sigs/jobset).
+Noone knows the job's name until it's created from JobSet, but each Job and Pod will get [`job-name`](https://github.com/kubernetes-sigs/jobset/blob/749a77c2647ac46b309b8227f66293e4116b97ed/api/v1alpha1/jobset_types.go#L26) label on it.
+
+Users can use `job-name` label to ensure that the Pods with the same `job-name` will land on the same rack 
+and Pods with other `job-name` won't labd on the same rack via `matchLabelSelectors`.
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha1
+kind: JobSet
+metadata:
+  name: pytorch
+  annotations:
+    alpha.jobset.sigs.k8s.io/exclusive-topology: rack
+spec:
+  replicatedJobs:
+    - name: workers
+      template:
+      ...
+        affinity:
+          podAffinity:      # ensures the pods of this job land on the same rack
+              requiredDuringSchedulingIgnoredDuringExecution:
+              - matchLabelSelectors:
+                  - matchLabelKey: job-name
+                    operator: In
+                topologyKey: rack
+            podAntiAffinity: # ensures only this job lands on the rack
+              requiredDuringSchedulingIgnoredDuringExecution:
+              - matchLabelSelectors:
+                  - matchLabelKey: job-name
+                    operator: NotIn
```

Comment:
Could you provide how your proposal looks via example yaml or something?

Reply:
Here are some examples:

#### a set of Pods A doesn't want to co-exist with other set of Pods, but want the set of Pods A co-located

```yaml
    spec:
      affinity:
        podAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: pod-set
                operator: ExistsWithSameValue
            topologyKey: kubernetes.io/hostname
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: pod-set
                operator: ExistsWithDifferentValue
            topologyKey: kubernetes.io/hostname
```

- smooth rolling upgrade for PodAntiAffinity:
```yaml
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values:
                - pause
            topologyKey: kubernetes.io/hostname
          - labelSelector:
              matchExpressions:
              - key: pod-template-hash
                operator: ExistsWithSameValue
            topologyKey: kubernetes.io/hostname
```

Reply:
> @Huang-Wei I am also not following, which operator are you referring to?

(The term "operator" is misleading...) I meant [LabelSelectorOperator](https://github.com/kubernetes/kubernetes/blob/ea974280dc9948e31d426ee6366abff0957a5f82/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1215). And see ^^ for examples.

Reply:
I like @Huang-Wei's proposal better

Reply:
> Specifically, how would ExistsWithSameValue/ExistsWithDifferentValue be interpreted in the absence of an object associated with the selector?

It should be errored out, or downgraded to the `Exists` semantics. As `ExistsWithSameValue/ExistsWithDifferentValue` relies on the presence of the key in the incoming Pod (or other API object).

> we need a new match function to pass the label value I guess, unless we inject it in the value field, but who will do that?

It'd be the concrete feature impl. (in this proposal it's scheduling). Meanwhile the generic match function can enforce the semantics of `Exists`.

Reply:
> It should be errored out, or downgraded to the Exists semantics.

But where do we error out? perhaps in `LabelSelectorAsSelector`/`LabelSelectorAsMap` functions? 

> It'd be the concrete feature impl. (in this proposal it's scheduling). Meanwhile the generic match function can enforce the semantics of Exists.

I guess `LabelSelectorAsSelector`/`LabelSelectorAsMap` functions will need to be extended to allow passing in the value (or create new ones that accept it); `ExistsWithSameValue` translates to "In value" (just like the original MatchLabelKeys semantics); while ExistsWithDifferentValue translates to two requirements `Key exists` and `NotIn value`, with this, the Match func implementation doesn't need to change.

>  or downgraded to the Exists semantics.
> Meanwhile the generic match function can enforce the semantics of Exists.

I am not sure about this implicit behavior, it can be surprising.

Reply:
@liggitt any thoughts or reservations to extending the operator scope in label selector?

Reply:
Generally adding a new enum to a set is problematic in the API.  The readers in this case are the scheduler and any external management tools (descheduler, etc).  Does anyone else read this field for enforcement?  The kubelet?

Reply:
> Does anyone else read this field for enforcement? The kubelet?

no one else in k/k will read it. (As you said, some external tools like descheduler or cluster autoscaler will read it though)

Reply:
but also, this is generally consumed via the labels pkg: https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apimachinery/pkg/labels; so if the new enums are handled there, most readers will just get them.

Reply:
Can you list the enum alternative in the "Alternatives" section of the KEP?

Reply:
Sure, added this idea as an alternative for now. 
Who can give the final decision on whether the enum addition is allowed or not? (API reviewer?)